### PR TITLE
Timeout and unit tests for agent healthcheck

### PR DESCRIPTION
**Pull request by sparrc**
### Summary

The docker healthcheck default timeout is 30s. We don't currently set any http timeouts on our internal healthcheck so if the request hangs then we don't log anything in agent or explicitly return a failed healthcheck to docker.

Adding a timeout of 25s means we can timeout in the agent codebase, log the error, and then explicitly return the failed healthcheck error code.

Also did a little refactor to allow adding unit tests.

New tests cover the changes: yes

test output:

```
% go test ./agent/app/. -count=1 -tags unit -v -run=Healthcheck
=== RUN   TestHealthcheck_Sunny
--- PASS: TestHealthcheck_Sunny (0.00s)
=== RUN   TestHealthcheck_InvalidURL2
--- PASS: TestHealthcheck_InvalidURL2 (0.00s)
=== RUN   TestHealthcheck_Timeout
level=error time=2020-04-23T21:51:23Z msg="error creating healthcheck request: parse   http://foobar: first path segment in URL cannot contain colon" module=healthcheck.go
level=error time=2020-04-23T21:51:25Z msg="health check [GET http://127.0.0.1:57227] failed with error: Get http://127.0.0.1:57227: net/http: request canceled (Client.Timeout exceeded while awaiting headers)" module=healthcheck.go
--- PASS: TestHealthcheck_Timeout (2.00s)
=== RUN   TestHealthcheck_404
--- PASS: TestHealthcheck_404 (0.00s)
PASS
ok  	github.com/aws/amazon-ecs-agent/agent/app	2.147s
```

### Description for the changelog

Log agent healthcheck timeouts.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/app/healthcheck_test.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestHealthcheck_Sunny(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (petderek) [inline agent/app/healthcheck_test.go:27]: Neat!

File: agent/app/healthcheck_test.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		fmt.Fprintln(w, "Hello, client")
+	}))
+	defer ts.Close()
+
+	rc := runHealthcheck(ts.URL, time.Second*2)
+	require.Equal(t, 0, rc)
+}
+
+func TestHealthcheck_InvalidURL2(t *testing.T) {
+	// leading space in url is invalid
```
Comment (petderek) [inline agent/app/healthcheck_test.go:37]: I know it’s supposed to be there but it still really bothers me.

File: agent/tcs/handler/handler.go [hunk 3]
```diff
@@ -87,6 +87,11 @@ func StartSession(params *TelemetrySessionParams, statsEngine stats.Engine) erro
 			seelog.Errorf("Error: lost websocket connection with ECS Telemetry service (TCS): %v", tcsError)
 			params.time().Sleep(backoff.Duration())
 		}
+		select {
```
Comment (sparrc) [inline agent/tcs/handler/handler.go:90]: Unrelated to this change, but this is a fix for when TestDoStartCgroupInitHappyPath has a failure after the test goroutine has already exited.

File: agent/engine/engine_unix_integ_test.go [hunk 4]
```diff
@@ -172,6 +172,7 @@ func createVolumeTask(scope, arn, volume string, autoprovision bool) (*apitask.T
 		DriverOpts: map[string]string{
 			"device": tmpDirectory,
 			"o":      "bind",
+			"type":   "tmpfs",
```
Comment (sparrc) [inline agent/engine/engine_unix_integ_test.go:175]: this was added for compatibility with newer docker versions, see https://github.com/docker/engine/commit/d5b271c155926057d33c2128ff57e49274455813

**Pull request merged**